hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/opt/WatermarkKeysAssigner.java

```diff
@@ -0,0 +1,284 @@
+    // Note: at the moment, no need to separate watermark keys without stream-to-stream join introduced.
+    private final byte keyCounter = 0;
+//    private final byte[] keyCounter = {0};
+
+    public WatermarkKeysAssigner(PhysicalRel root) {
+        this.root = root;
+        this.visitor = new BottomUpWatermarkKeyAssignerVisitor();
+    }
+
+    public void assignWatermarkKeys() {
+        visitor.go(root);
+    }
+
+    public Map<Integer, Byte> getWatermarkedFieldsKey(RelNode node) {
+        return visitor.getRelToWmKeyMapping().get(node);
+    }
+
+    /**
+     * Main watermark key assigner used to propagate keys
+     * from bottom to top over whole rel tree.
+     */
+    private class BottomUpWatermarkKeyAssignerVisitor extends RelVisitor {
+        private final Map<RelNode, Map<Integer, Byte>> relToWmKeyMapping = new HashMap<>();
+
+        BottomUpWatermarkKeyAssignerVisitor() {
+        }
+
+        public Map<RelNode, Map<Integer, Byte>> getRelToWmKeyMapping() {
+            return relToWmKeyMapping;
+        }
+
+        @Override
+        public void visit(RelNode node, int ordinal, @Nullable RelNode parent) {
+            // front wave of recursion
+
+            if (node instanceof FullScanPhysicalRel) {
+                assert node.getInputs().isEmpty() : "FullScan is not a leaf";
+                FullScanPhysicalRel scan = (FullScanPhysicalRel) node;
+                int idx = scan.watermarkedColumnIndex();
+                if (idx >= 0) {
+                    scan.setWatermarkKey(WatermarkKeysAssigner.this.keyCounter);
+                    relToWmKeyMapping.put(scan, Collections.singletonMap(idx, keyCounter));
+                }
+                return;
+            }
+
+            super.visit(node, ordinal, parent);
+
+            // back wave of recursion
+            if (node instanceof CalcPhysicalRel) {
+                CalcPhysicalRel calc = (CalcPhysicalRel) node;
+                List<RexNode> projects = calc.getProgram().expandList(calc.getProgram().getProjectList());
+                Map<Integer, Byte> refByteMap = relToWmKeyMapping.get(calc.getInput());
+                if (refByteMap == null) {
+                    return;
+                }
+
+                int calcIndex = 0;
+                Map<Integer, Byte> calcRefByteMap = new HashMap<>();
+                for (RexNode rexNode : projects) {
+                    if (rexNode instanceof RexInputRef) {
+                        int idx = ((RexInputRef) rexNode).getIndex();
+                        calcRefByteMap.put(calcIndex++, refByteMap.get(idx));
+                    }
+                }
+
+                relToWmKeyMapping.put(calc, calcRefByteMap);
+            } else if (node instanceof UnionPhysicalRel) {
+                UnionPhysicalRel union = (UnionPhysicalRel) node;
+                Set<Integer> usedColumns = new HashSet<>();
+                for (RelNode input : union.getInputs()) {
```

Comment:
We need to do intersection of watermarked fields. A field is watermarked if it's watermarked in all inputs.